lib/ts/controllers/s-modal.ts

```diff
@@ -153,58 +153,69 @@ namespace Stacks {
-            }
+        private lastVisible(elements: HTMLElement[]) {
+            return [...elements].reverse().find(el => el.offsetParent !== null);
+        }
 
-            // focus on the first focusable item within the modal
+        /**
+         * Binds tab presses on tabbable items such that tabbing only works within the modal
+         */
+        private focusInsideModal() {
             this.modalTarget.addEventListener("s-modal:shown", () => {
-                // double check the element still exists when the event is called
-                if (initialFocus && document.body.contains(initialFocus)) {
-                    initialFocus.focus()
-                }
+                var initialFocus = this.firstVisible(this.initialFocusTargets) ?? this.firstVisible(this.getAllTabbables());
+                initialFocus?.focus();
             }, {once: true });
+        }
 
-            var firstTabbable = <HTMLElement>allTabbables[0];
-            var lastTabbable = <HTMLElement>allTabbables[allTabbables.length - 1];
-
-            // if the first or last item is tabbed over, ensure that the focus "loops" back to the end of the array instead of leaving the modal
-            this._boundTabTrap = this._boundTabTrap || ((e: KeyboardEvent) => {
-                // if somehow the user has tabbed out of the modal or if focus started outside the modal, push them to the first item
-                if (!this.modalTarget.contains(<Element>e.target)) {
-                    e.preventDefault();
-                    firstTabbable.focus();
-                }
 
-                // if they've tabbed backwards over the first item, then go to the last item
-                if (e.target == firstTabbable && e.keyCode === 9 && e.shiftKey) {
-                    e.preventDefault();
-                    lastTabbable.focus();
-                }
+        /**
+         * Returns keyboard focus to the modal if it has left or is about to leave.
+         */
+        private keepFocusWithinModal(e: KeyboardEvent) {
 
-                // if they've tabbed forwards over the last item, then go to the first item
-                if (e.target == lastTabbable && e.keyCode === 9 && !e.shiftKey) {
+            // If somehow the user has tabbed out of the modal or if focus started outside the modal, push them to the first item.
+            if (!this.modalTarget.contains(<Element>e.target)) {
+                var focusTarget = this.firstVisible(this.getAllTabbables());
+                if (focusTarget) {
                     e.preventDefault();
-                    firstTabbable.focus();
+                    focusTarget.focus();
                 }
-            });
 
-            document.addEventListener("keydown", this._boundTabTrap);
+                return;
+            }
 
-            return initialFocus;
+            // If we observe a tab keydown and we're on an edge, cycle the focus to the other side.
+            if (e.keyCode === 9) {
```

Comment:
nit: Don't use `KeyboardEvent.keyCode`, it is [deprecated](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode). We can use the friendlier [`.key`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key) instead (IE9+):

```suggestion
            if (e.key === "Tab") {
```

I just realized that the code I wrote originally was using `.keyCode`, so... I dunno what I was thinking. Let's go ahead and fix it up here and brush my terrible code under the rug...